### PR TITLE
Calculation Endpoints: Adds Validation and Tests 

### LINKED ISSUE #1499: Engineer tests and validation for 'Calculation' Web API requests
### Overview

We need to create a workflow for testing all 'calculation' endpoints. We will need to set up tests and web API request validation for the endpoints.

### Details | Scope of Work
We need to test the API endpoints using 'jest'  

### Action Items

- [x] write all tests for endpoints and documentation
- [x] set up server side validation schema

### Resources/Instructions
Server Side Validation
- https://www.npmjs.com/package/express-json-validator-middleware
- https://simonplend.com/how-to-handle-request-validation-in-your-express-api/  

Testing
- https://jestjs.io/
- https://testcontainers.com/

### PR DESCRIPTION
**Pull request by agosmou**
Fixes #1499 

Sub-Issue from #1238 

### What changes did you make?

- Added integration tests for the 'calculation' endpoints that test the functional requirements of the calculation workflow as well as edge cases
- No need for validation as overexposed endpoints were deleted

### Why did you make the changes (we will use this info to test)?

- Calculation Workflow Quality: Added tests for 'calculation' endpoints to ensure workflow meets functional requirements and handles edge cases
- Enhanced CI/CD: Integration tests in TDM CI pipeline prevent PRs that merge breaking changes
- Reliable Testing: Integrated 'Jest' and 'testcontainers' to add robust testing reflecting our production setup

### Notes
See [Test Wiki](https://github.com/hackforla/tdm-calculator/wiki/Testing#testing-the-tdm-server) for testing documentation on TDM Calculator

### CODE REVIEW

File: server/.env.example
Comment (agosmou) [inline server/.env.example:1]: I changed the file name from 
.example.env -> .env.example  as it was before (better naming)

File: server/app/services/calculation.service.js
Comment (agosmou) [inline server/app/services/calculation.service.js:41]: This was returning an undefined variable so I used the VSCode debugger to check what the property was supposed to be - it's been rectified now and actually returns the `id`

File: server/__tests__/calculation.test.js
Comment (agosmou) [inline server/__tests__/calculation.test.js:110]: Consider adding issue - inexistent id's also returned 200 status codes

File: server/__tests__/calculation.test.js
Comment (agosmou) [inline server/__tests__/calculation.test.js:121]: Consider adding issue - inexistent id's also returned 200 status codes

File: server/app/services/calculation.service.js
Comment (ZekeAranyLucas) [reply server/app/services/calculation.service.js:41]: Nice!

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [reply server/__tests__/calculation.test.js:121]: Agreed.

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [reply server/__tests__/calculation.test.js:110]: Agreed

File: server/.env.example
Comment (ZekeAranyLucas) [reply server/.env.example:80]: Does this mean tdm-server-tests can run with .env.example?

File: server/app/schemas/calculation.post.js
Comment (ZekeAranyLucas) [inline server/app/schemas/calculation.post.js:8]: What will happen in the UX if the user breaks validation?

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [inline server/__tests__/calculation.test.js:92]: expect()?

Comment (agosmou) [reply server/__tests__/calculation.test.js:92]: line 92 intentional - this sets the variable that is used in the following tests

File: server/app/schemas/calculation.post.js
Comment (agosmou) [reply server/app/schemas/calculation.post.js:8]: I think the create, update , delete, calculation endpoints are [not used](https://github.com/search?q=repo%3Ahackforla%2Ftdm-calculator+calculation+path%3A%2F%5Eclient%5C%2Fsrc%5C%2F%2F&type=code&p=1) by the client currently - future implementation. I think we mostly use the GET endpoints. Also, there is only '1' calculation in the db currently.

File: server/.env.example
Comment (agosmou) [reply server/.env.example:80]: Nope! I just changed it because VSCode has better highlighting with the new name. Also included the variables so new members could run the tests easily without the formal variables. I tried to get this to work unsuccessfully with `./push.sh` 
An alternate strategy I thought of this morning is a function in the push script that can uncomment the test variables for the server tests and then re-comment upon completion. I'll have to check if this is viable, but editing the .env file with code feels unkosher. Thoughts?

File: server/app/schemas/calculation.post.js
Comment (ZekeAranyLucas) [reply server/app/schemas/calculation.post.js:8]: Then how do you know if this validation is right?

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [reply server/__tests__/calculation.test.js:92]: Other tests depend on this side effect? Tests need to runnable as a single unit.

File: server/app/schemas/calculation.post.js
Comment (ZekeAranyLucas) [reply server/app/schemas/calculation.post.js:8]: The answer could be you are making stubs, but then the code needs a comment that the regex are not spec'ed. 

I am looking for answers to why in the code.

Comment (ZekeAranyLucas) [reply server/app/schemas/calculation.post.js:8]: And unused endpoints are risky. I am inclined to delete dead code as a security risk. 

Is there a plan for when to add it?

File: server/.env.example
Comment (agosmou) [reply server/.env.example:80]: I was looking into that but had some concerns.

Since the container has a single database that is [restored before every test suite](https://github.com/hackforla/tdm-calculator/blob/develop/server/_jest-setup_/local-setup/setup-after-env.js) (i.e. before every test file), the logic that handles this will call `tedious-pool.js` the [connection logic](https://github.com/hackforla/tdm-calculator/blob/develop/server/app/services/tedious-pool.js) from the server which will consume the original .env file. this means that this file would require a conditional to see which `.env` to consume right? Is this crufty?

File: server/__tests__/calculation.test.js
Comment (agosmou) [reply server/__tests__/calculation.test.js:92]: It's intentional for the nature of these tests since it avoids mocking - everything tries to be as "real" as possible. We discussed adding functional tests that would be as close to the production environment as possible. So certain tests have protected endpoints which require JWT's

We had discussed trying to add unit tests but this way added more coverage and better ROI on time for where we are in the project.

Thoughts?

File: server/app/schemas/calculation.post.js
Comment (agosmou) [reply server/app/schemas/calculation.post.js:8]: excellent question. Ill have to refer to @entrotech

not to mention validation logic may even require numbers to be valid (not included currently)

**Comment by agosmou** [comment]
> Looking good. What prompted adding the schema validation?

@ZekeAranyLucas 

Request from John to add validation for all web api endpoints - #1238

File: server/.env.example
Comment (ZekeAranyLucas) [reply server/.env.example:80]: It means everywhere we use dotenv should probably go through a tdmenv wrapper.

Comment (agosmou) [reply server/.env.example:80]: so new issue for refactoring codebase with this ^

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [reply server/__tests__/calculation.test.js:92]: For example, jest can rerun just failed tests or based on regex. As your tests scale, running subsets and reordering will become even more important. 

Don't let tests depend on other tests.

**Review by ZekeAranyLucas — CHANGES_REQUESTED** [review]
In the name of security, I recommend deleting unused service  methods. This change becomes much smaller. 😀

File: server/__tests__/calculation.test.js
Comment (agosmou) [reply server/__tests__/calculation.test.js:92]: > If the tests must run in a certain order they should just be a single test.
I put it in a single 'describe' block with this intention - is that pattern inappropriate? 

> Don't let tests depend on other tests.
Could this be done without mocking?

**Comment by ZekeAranyLucas** [comment]
> In the name of security, I recommend deleting unused service  methods. This change becomes much smaller. 😀 

After looking at #1238, I have better context. Write tests for the faq crud APIs first. Those ones are accessible in the client, and need the attention.

**Comment by ZekeAranyLucas** [comment]
Ok, editing calculations is covered by issuer #46. Given the nature of how rules work, I suggest only devs should ever change rules. They are too finicky for normal users to get right. Allowing an admin to delete a rule is a recipe for disaster

I recommend:

* resolve 46 as "won't fix"
* delete related service methods
* use #1238 issue to identify similar overexposed public surface and delete.

What do you think @entrotech ?

File: .github/workflows/tdm-server-tests.yml
Comment (agosmou) [inline .github/workflows/tdm-server-tests.yml:8]: better naming convention for the workflow

File: server/__tests__/calculation.test.js
Comment (ZekeAranyLucas) [reply server/__tests__/calculation.test.js:92]: Describe is any arbitrary semantic group. "It" is the test. 

So you could group tests by all needing the same before/after assumptions. 

```
describe ('tests that need an calculation')
  let calcId = null
  beforeEach() { calcId = ... }
  afterEach() { cleanup(calcId)}
  it('verifies delete') { 
    delete(calcId);
    expect(get(calcId)).toBeFalsy()
  }
  it('verifiers update') ...
```

So each test starts with clean assumptions.

Comment (agosmou) [reply server/__tests__/calculation.test.js]: Just refactored based on this suggestion

commit 7f1deff8af2698b17763ceb17ab8f937677fc188

**Comment by ZekeAranyLucas** [comment]
@agosmou is this ready to roll? are you going to create issues after merging?

**Comment by entrotech** [comment]
@ZekeAranyLucas  comments above are spot-on. Early on, we had thoughts of implemented admin screens for versioning the calculations, allowing a super-admin to edit rules on an existing calculation to create a new version, and apply that version to future projects, but the implications are kinda scary, and I like having enough friction in changing rules that they are made mostly by DB Migrations and an occasional code change, so they need to go through a QA process by people that know what they are doing.

**Review by ZekeAranyLucas — APPROVED** [review]

**Comment by agosmou** [comment]
The ['rule service'](https://github.com/hackforla/tdm-calculator/blob/develop/client/src/services/rule.service.js#L16) uses the getById service from the 'calculations' server service so this will remain for now.

**Comment by agosmou** [comment]
The above requested changes were addressed in the latest commit [f1dcd6d](https://github.com/hackforla/tdm-calculator/pull/1521/commits/f1dcd6d232b5387b00f7983914222fa2e4c0e9b9)

**Review by entrotech — APPROVED** [review]
I trust you two to have worked out the details on this PR. I am agreeing on deleting the endpoints you mentioned.

BTW, there is a remnant of Issue #46 that I can't bring myself to remove, just cause it's kinda cool.  Log in as an admin and change the client path to /admin and it give you an interesting view of the rules that make up the calculation.

@agosmou you can merge this yourself, if you want.

**Comment by ZekeAranyLucas** [comment]
/admin is cool!  i wouldn't remove that.

**Comment by agosmou** [comment]
Yes! Very cool (see below for record). This is unaffected by this PR.  

---
image

**Pull request merged**
